Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+/*
+ * This file contains the optimizer for CPython uops.
+ * It performs a traditional data-flow analysis[1] over the trace of uops.
+ * Using the information gained, it chooses to emit, or skip certain instructions
+ * if possible.
+ *
+ * [1] For information on data-flow analysis, please see page 27 onwards in
+ * https://ilyasergey.net/CS4212/_static/lectures/PLDI-Week-12-dataflow.pdf
+ * Credits to the courses UPenn Compilers (CIS 341) and NUS Compiler Design (CS4212).
+ *
+ * */
 #include "Python.h"
-#include "opcode.h"
 #include "pycore_interp.h"
 #include "pycore_opcode_metadata.h"
 #include "pycore_opcode_utils.h"
 #include "pycore_pystate.h"       // _PyInterpreterState_GET()
 #include "pycore_uop_metadata.h"
 #include "pycore_long.h"
 #include "cpython/optimizer.h"
+#include "pycore_optimizer.h"
+#include "pycore_object.h"
+#include "pycore_dict.h"
+#include "pycore_function.h"
+#include "pycore_uop_metadata.h"
+#include "pycore_uop_ids.h"
+#include "pycore_range.h"
+
+#include <stdarg.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stddef.h>
-#include "pycore_optimizer.h"
+
+#define MAX_ABSTRACT_INTERP_SIZE 2048
+
+#define OVERALLOCATE_FACTOR 5
+
+#define PEEPHOLE_MAX_ATTEMPTS 3
+
+// + buffer to account for implicit root frame.
+#define MAX_ABSTRACT_FRAME_DEPTH (TRACE_STACK_SIZE + 2)
```

Comment:
I am trrying this with +1 instead of +2, and I'm not getting any crashes. Maybe that was in an earlier version you tried?

```suggestion
#define MAX_ABSTRACT_FRAME_DEPTH (TRACE_STACK_SIZE + 1)
```

Reply:
Oh wait, I get *one* crash (in `test.test_tracemalloc.TestSnapshot.test_filter_traces_domain_filter`). Let me debug that.

Reply:
Hah, found it. We do need two extra. The reason is that when the trace construction detects the stack overflow (in the `TRACE_STACK_PUSH()` macro), it has already emitted the `_PUSH_FRAME` uop. At that point it just emits `_EXIT_TRACE` and continues, so we will still attempt to optimize the trace. I guess it's worth a comment here.

(Alternatively, you could check for `operand == NULL` -- since in that case the `operand` field is not set yet.)
```suggestion
// Need extras for root frame and for overflow frame (see TRACE_STACK_PUSH())
#define MAX_ABSTRACT_FRAME_DEPTH (TRACE_STACK_SIZE + 2)
```